Add rule to check if a footer line is in body

parhamsaremi opened the pull request:
Messages that start with "Fixes" or "[]" should be in the footer this newly added rule will check this behavior and will raise errors if it's not correctly addressed.

knocte commented on commitlintplugins.test.ts:
```diff
@@ -181,6 +181,38 @@ test('empty-wip-3', () => {
+        + "\n\n"+ "Bla bla blah[1]." 
+        + "\n\n" + "[1] http://foo.bar/baz";
+    let footerCorrectPlace2 = runCommitLintOnMsg(commitMsgWithWrongFooter);
+    console.log('HERE ============>' + footerCorrectPlace2.stdout)
```
@parhamsaremi please remove logging (or comment it)

knocte commented on commitlint.config.ts:
```diff
@@ -167,6 +173,34 @@ module.exports = {
                     ]
                 },
 
+                'footer-correct-place': ({body}: {body:any}) => {
```
@parhamsaremi I'm not convinced about the name of the rule, let's rather name it 'footer-notes-misplacement' (and as a consequence let's rename some vars in the tests)

knocte commented on commitlint.config.ts:
```diff
@@ -167,6 +173,34 @@ module.exports = {
+                        let seenFooter = false;
+                        let lines = bodyStr.split(/\r?\n/);
+                        for (let line of lines) {
+                            if (isEmptyLine(line)){
```
@parhamsaremi do we really need this check? let's think about it, would an empty line return true for the function isFooterNote? I guess not, so I don't think we need this extra code (maybe I'm wrong, please double check)

parhamsaremi replied:
Yes. an empty line won't be considered as footer but !isFooterNote will be true which is not the behavior I want so I skipped empty lines :thinking: (I also commented out this check and one test failed)

knocte replied:
ah fair enough, my bad

knocte replied:
but creating a function for checking if .length is zero looks like overengineering to me :)

knocte replied:
you're not using that function anywhere else... just check if length is zero right there

parhamsaremi replied:
hmmm sure I'll fix it right now.

knocte commented:
@parhamsaremi I think you learned a bad habit from our propensity to ask for squashed branches in geewallet/LN work. Squashing branches is only needed when there are WIP commits (and only if there's more than 1 WIP commit in certain branch). In this case, you worked on this PR for only one day so I believe there wouldn't be any WIP commit. What I was expecting is to see 1 commit first where you add the tests (and CI failing because of that), then a commit where you add the rule and makes the test pass. Please apply TDD as much as you can (and that includes: when using with git).

knocte commented on commitlint.config.ts:
```diff
@@ -167,6 +173,34 @@ module.exports = {
+                    }
+                    return [
+                        !offence,
+                        `Please add footer messages after body paragraphs`
```
@parhamsaremi this error message starting with "Please add" would make a developer think that he has to "add stuff" (as in, make the commit message bigger), but this is not what this rule should advice, the thing to be done is to shuffle the content around, so let's not use the word "add" here, it's confusing

parhamsaremi commented:
> @parhamsaremi I think you learned a bad habit from our propensity to ask for squashed branches in geewallet/LN work. Squashing branches is only needed when there are WIP commits (and only if there's more than 1 WIP commit in certain branch). In this case, you worked on this PR for only one day so I believe there wouldn't be any WIP commit. What I was expecting is to see 1 commit first where you add the tests (and CI failing because of that), then a commit where you add the rule and makes the test pass. Please apply TDD as much as you can (and that includes: when using with git).

Based on our talk last night I learned what I did wrong and I'll work with this standard in future works. Should I split my squashed commit into two new commits in this PR? :thinking:

knocte commented:
> Should I split my squashed commit into two new commits in this PR?

No need, that was a comment to take in account for future, not an actionable comment. That being said, this PR needs to be rebased sorry.

parhamsaremi commented:
> > Should I split my squashed commit into two new commits in this PR?
> 
> No need, that was a comment to take in account for future, not an actionable comment. That being said, this PR needs to be rebased sorry.

Yup saw it after addressed the comments and pushed again. it should be ok now.

parhamsaremi commented:
@knocte It needed another rebase. now everything should be ok :thinking:
